jq255.md

```diff
@@ -0,0 +1,1425 @@
+and party 2) generate private/public key pairs (sk1 and Q1 for party 1,
+sk2 and Q2 for party 2). Each party receives a copy of the public key of
+the other party, and combines it with its own private key to produce a
+secret key (of size 32 bytes). If the public keys were computed
+correctly and not altered in transit, then both parties compute the same
+secret key, and that key is not known to outsiders. It can then be used
+for further cryptographic operations, typically symmetric encryption.
+
+**Warning:** Two important caveats apply:
+
+  - The key exchange described here does not, by itself, ensure
+    authentication. An active attacker in position to intercept messages
+    exchanged between the two parties may replace public keys on the
+    wire with values generated by the attacker, and establish distinct
+    secret keys, but both known to the attacker, with each party. The
+    attacker may then decrypt and re-encrypt subsequent messages as they
+    flow. A Diffie-Hellman key exchange may provide security against
+    active attackers only if integrated into an outer protocol that
+    ensures proper authentication (e.g. SSL/TLS uses signatures and
+    certificates on top of a Diffie-Hellman key exchange).
+
+  - While private and public keys used here have the same format as
+    private and public keys used for signatures, it is **not
+    recommended** to use the same private/public key pair for both key
+    exchange and signatures. In general, the two activities call for key
+    pairs with distinct, incompatible lifecycle properties (simply put,
+    keys used for encryption should normally be backed up to prevent
+    data loss in case of key loss, while keys used for signatures must
+    not be backed up, since their value relies in their exclusive
+    control by the purported signer). Moreover, possible interactions
+    between the two schemes are not entirely explored.
+
+The process below describes how a party uses its own key pair (`sk` and
+`Q`) along with the *encoded* public key received from the peer
+(`peer_pk`) to compute the shared secret. The function returns two
+values: the computed secret key (presumably shared with the peer), and a
+Boolean status. The status is set to `False` if the bytes sent by the
+peer cannot be decoded as a proper public key. When the peer bytes are
+invalid, a secret key is still computed, and it is not guessable by
+outsiders. The point of producing such a fake secret key is to allow the
+use of the key exchange in some (rather contrived) scenarios in which
+attackers cannot *see*, but can *alter* public keys in transit, and may
+obtain some extra information from knowledge of whether the altered key
+was valid or not. Computing a fake but indistinguishable secret key can
+provide this extra security feature. In most normal uses of
+Diffie-Hellman, that feature is not needed.
+
+```python
+def ECDH(sk, Q, peer_pk):
```

Comment:
is this 'ecdh' or 'just' DH when over prime-order groups that just happen to use elliptic curves under the hood

Reply:
Purists would say that Diffie and Hellman only talked about the multiplicative subgroup of integers modulo a big prime, and that anything else is a different algorithm. I am using here "ECDH" mostly because it's an already known acronym with a lot of visibility (e.g. you see it in the names of TLS cipher suites) and it is not, technically, _wrong_ in our case, because this is really done over an elliptic curve; the only difference with the "usual" ECDH is in how the resulting shared point is turned into a shared key (plain ECDH uses the x coordinate of the point, so it maps P and -P to the same output; here, the group encoding is used, so P and -P are mapped to distinct outputs, but, _a contrario_, P and P+N are mapped to the same value).

Reply:
@FiloSottile @str4d any opinions on this naming or am I just thinking too hard about this